src/main/java/com/amazonaws/encryptionsdk/kms/KmsClientSupplier.java

```diff
@@ -0,0 +1,212 @@
+                    kmsClientBuilder = kmsClientBuilder.withRegion(regionId);
+                }
+
+                AWSKMS client = kmsClientBuilder.build();
+
+                if (clientCachingEnabled) {
+                    client = newCachingProxy(client, regionId);
+                }
+
+                return client;
+            };
+        }
+
+        /**
+         * Sets the AWSCredentialsProvider used by the client.
+         *
+         * @param credentialsProvider New AWSCredentialsProvider to use.
+         */
+        public Builder credentialsProvider(AWSCredentialsProvider credentialsProvider) {
+            this.credentialsProvider = credentialsProvider;
+            return this;
+        }
+
+        /**
+         * Sets the ClientConfiguration to be used by the client.
+         *
+         * @param clientConfiguration Custom configuration to use.
+         */
+        public Builder clientConfiguration(ClientConfiguration clientConfiguration) {
+            this.clientConfiguration = clientConfiguration;
+            return this;
+        }
+
+        /**
+         * Sets the AWS regions that the client supplier is permitted to use.
+         *
+         * @param regions The set of regions.
+         */
+        public Builder allowedRegions(Set<String> regions) {
+            notEmpty(regions, "At least one region is required");
+            this.allowedRegions = Collections.unmodifiableSet(new HashSet<>(regions));
+            return this;
+        }
+
+        /**
+         * Sets the AWS regions that the client supplier is not permitted to use.
+         *
+         * @param regions The set of regions.
+         */
+        public Builder excludedRegions(Set<String> regions) {
+            requireNonNull(regions, "regions is required");
+            this.excludedRegions = Collections.unmodifiableSet(new HashSet<>(regions));
+            return this;
+        }
+
+        /**
+         * When set to true, allows for the AWSKMS client for each region to be cached and reused.
+         *
+         * @param enabled Whether or not caching is enabled.
+         */
+        public Builder clientCaching(boolean enabled) {
+            this.clientCachingEnabled = enabled;
+            return this;
+        }
+
+        /**
+         * Creates a proxy for the AWSKMS client that will populate the client into the client cache
+         * after a KMS method successfully completes or a KMS exception occurs. This is to prevent a
+         * a malicious user from causing a local resource DOS by sending ciphertext with a large number
+         * of spurious regions, thereby filling the cache with regions and exhausting resources.
+         *
+         * @param client   The client to proxy
+         * @param regionId The region the client is associated with
+         * @return The proxy
+         */
+        private AWSKMS newCachingProxy(AWSKMS client, String regionId) {
+            return (AWSKMS) Proxy.newProxyInstance(
```

Comment:
It's worth keeping an eye on this. Proxies can cause performance problems sometimes as they depend on Java reflection APIs.

Reply:
Sure, though I hope any proxy overhead would be insignificant compared to the actual service call to KMS.